db/migrate/20200709133910_add_linelist_fields_to_patients.rb

```diff
@@ -0,0 +1,84 @@
+class AddLinelistFieldsToPatients < ActiveRecord::Migration[6.0]
+  def up
+    add_column :patients, :latest_assessment_at, :datetime, index: true
+    add_column :patients, :latest_fever_or_fever_reducer_at, :datetime, index: true
+    add_column :patients, :latest_positive_lab_at, :date, index: true
+    add_column :patients, :negative_lab_count, :integer, default: 0, index: true
+    add_column :patients, :latest_transfer_at, :datetime, index: true
+    add_column :patients, :latest_transfer_from, :integer, index: true
+
+    # populate :latest_assessment_at
+    execute <<-SQL.squish
+      UPDATE patients
+      INNER JOIN (
+        SELECT patient_id, MAX(created_at) AS latest_assessment_at
+        FROM assessments
+        GROUP BY patient_id
+      ) t ON patients.id = t.patient_id
+      SET patients.latest_assessment_at = t.latest_assessment_at  
+    SQL
+
+    # populate :latest_fever_or_fever_reducer_at
+    execute <<-SQL.squish
+      UPDATE patients
+      INNER JOIN (
+        SELECT assessments.patient_id, MAX(assessments.created_at) AS latest_fever_or_fever_reducer_at
+        FROM assessments
+        INNER JOIN conditions ON assessments.id = conditions.assessment_id
+        INNER JOIN symptoms ON conditions.id = symptoms.condition_id
+        WHERE (symptoms.name = 'fever' OR symptoms.name = 'used-a-fever-reducer') AND symptoms.bool_value = true
+        GROUP BY assessments.patient_id
+      ) t ON patients.id = t.patient_id
+      SET patients.latest_fever_or_fever_reducer_at = t.latest_fever_or_fever_reducer_at
+    SQL
+
+    # populate :latest_positive_lab_at
+    execute <<-SQL.squish
+      UPDATE patients
+      INNER JOIN (
+        SELECT patient_id, MAX(specimen_collection) AS latest_positive_lab_at
+        FROM laboratories
+        WHERE result = 'positive'
+        GROUP BY patient_id
+      ) t ON patients.id = t.patient_id
+      SET patients.latest_positive_lab_at = t.latest_positive_lab_at
+    SQL
+
+    # populate :negative_lab_count
+    execute <<-SQL.squish
+      UPDATE patients
+      INNER JOIN (
+        SELECT patient_id, COUNT(*) AS negative_lab_count
+        FROM laboratories
+        WHERE result = 'negative'
+        GROUP BY patient_id
+      ) t ON patients.id = t.patient_id
+      SET patients.negative_lab_count = t.negative_lab_count
+    SQL
+
+    # populate :latest_transfer_at and :latest_transfer_from
+    execute <<-SQL.squish
```

Comment:
Are these more efficient being raw SQL? Just curious about why you went this route.

Reply:
When I was looking up how other people executed queries along with their migrations, it seemed like I was seeing more examples with people using raw SQL. Here's one of the articles I skimmed over: https://makandracards.com/makandra/15575-how-to-write-complex-migrations-in-rails

I think raw SQL is probably faster/more efficient than active record, especially for big queries like this. When I was testing this migration out on my local machine btw, it only took about a minute to finish all of them on ~80k patients.

Reply:
Excellent, and makes perfect sense to me.